drivers/ds3231: add alarm support

nicoHarel opened the pull request:
### Contribution description

See also issue #16155. 

This PR adds support for alarms within the DS3231 drivers. 

I have corrected some alarm registers which were not used before and seemed to be incorrect. 

This is still a work in progress:
- As mentioned in issue #16155 it might be better to hide the alarm functions behind a pseudo module even though, as I mentioned earlier, I do not completely agree with this. You can absolutely use RTC alarms without any kind of callback. Sometimes the alarms are not even intended to the MCU. Meaning in those cases there is no need for a callback.
- I used typedefs for the alarm configuration, it seemed way easier/cleaner to me, but others might disagree/have a better idea. 
- The ITCN bit is for now set during the initialization of the RTC, therefore disabling the square wave output (it was already disabled I think). It might be better to create some kind of options in case someone wants to use the SQW output. Not sure if that is important. 
- ~~This is a basic alarm implementation, I didn't provide anything to read the alarm registers, just the configuration part.~~
- ~~The doxygen documentation for the added function in include/ds3231.h has not been written yet since there might be some modifications.~~

### Testing procedure
I did not modify the tests for the DS3231 driver. I've tested the code myself. I don't know how to test an alarm that would normally trigger for example the next day.  Maybe we can only test a few settings (alarms that would trigger in the next few seconds) ? 

Have a good day.

nicoHarel commented:
> > Maybe we can only test a few settings (alarms that would trigger in the next few seconds) ?
> 
> The test application code can be modified to be tested like the [periph_rtc](https://github.com/RIOT-OS/RIOT/blob/master/tests/periph_rtc/main.c#L16-L17) testing is done, the alarm is set in an interval of 10 s.
Good for me.

kaspar030 commented:
> The test application code can be modified to be tested like the [periph_rtc](https://github.com/RIOT-OS/RIOT/blob/master/tests/periph_rtc/main.c#L16-L17) testing is done, the alarm is set in an interval of 10 s.
> Good for me.

can this be reduced to e.g., two seconds? It's nice to have tests complete quickly (as in, only run as long as necessary to test what is being tested).

nicoHarel commented:
Added an alarm test (2 seconds delay) after the time test. Is this enough ? 

I also added two functions in the driver to get the alarm flag state inside the status register. 

I did not provide a shell command to test the alarm individually. Should I ?

benpicco commented:
Shouldn't the RTC alarm cause some GPIO pin to toggle, so we can get an interrupt?
This way the RTC can be used to wake the CPU.

Is it correct that the alarm can only be triggered for a certain time, not a date?

nicoHarel commented:
> Shouldn't the RTC alarm cause some GPIO pin to toggle, so we can get an interrupt?
> This way the RTC can be used to wake the CPU.

This is the question I had in my first post on this PR and in the discussion we had with @jeandudey in issue #16155. What you are mentioning is a specific use case of an RTC alarm in my opinion (although the most common one probably). But you can also use the alarm to trigger something else: wake-up of external components, trigger external events. I use an NRF52 meaning it can't even wake up from deep-sleep with a GPIO interrupt in RIOT, if I'm not mistaken. 

I think (giving that you and @jeandudey seem to agree on this) that the way to go is more or less the solution mentioned in issue #16155:
- Look for a callback in function argument => create GPIO interrupt => call the callback function when interrupt triggered. 
- If no callback provided: don't try to create a GPIO interrupt. 
- If no GPIO interrupt support: don't look for a callback.  (I don't really know how to do this, I'm guessing some kind of pseudo-module as @jeandudey mentioned). 

Don't know if that's clear, but if it is, do we all agree on this ? 

I very much want to have the possibility to not provide a callback nor a GPIO for the interrupt, I don't want to waste a GPIO if I don't need it or create a callback function that would do just nothing. 

> 
> Is it correct that the alarm can only be triggered for a certain time, not a date?
 
No it is not correct, I forgot to provide one more value in the enumerates (case when A1M4 A1M3 A1M2 A1M1 are all equal to 0, same for A2MX registers for alarm 2). Thank you for pointing that out.

jeandudey commented on drivers/ds3231/ds3231.c:
```diff
@@ -236,6 +239,140 @@ int ds3231_set_time(const ds3231_t *dev, const struct tm *time)
+    }
+}
+
+int ds3231_toggle_alarm_2(const ds3231_t *dev, bool state)
```
Same as before

jeandudey commented on drivers/ds3231/ds3231.c:
```diff
@@ -236,6 +239,140 @@ int ds3231_set_time(const ds3231_t *dev, const struct tm *time)
+
+}
+
+int ds3231_get_alarm_1_flag(const ds3231_t *dev, bool* flag)
```
```suggestion
int ds3231_get_alarm_1_flag(const ds3231_t *dev, bool *flag)
```

jeandudey commented on drivers/ds3231/ds3231.c:
```diff
@@ -236,6 +239,140 @@ int ds3231_set_time(const ds3231_t *dev, const struct tm *time)
+        return -EIO;
+    }
+
+    *flag = (raw & STAT_A1F) ;
```
```suggestion
    *flag = (raw & STAT_A1F);
```

As pointed by the Github Actions check (can be seen on the actions result or the files view as a suggestion 😉 )

jeandudey commented on drivers/ds3231/ds3231.c:
```diff
@@ -236,6 +239,140 @@ int ds3231_set_time(const ds3231_t *dev, const struct tm *time)
+        return -EIO;
+    }
+
+    *flag = (raw & STAT_A2F) >> 1 ;
```
Same as before

jeandudey commented on drivers/ds3231/ds3231.c:
```diff
@@ -236,6 +239,140 @@ int ds3231_set_time(const ds3231_t *dev, const struct tm *time)
+    uint8_t a1mxMask[A1_REG_NUMOF];
+
+    /* A1M1, A1M2, A1M3 and A1M4 are set accordingly to the trigger type */
+    a1mxMask[0] = (trigger & 0x01) << 7;
```
Please use lowercase name for variables, e.g.: a1mx_mask 😉

nicoHarel replied:
Yes sorry, force of habit :grimacing:

jeandudey reviewed (changes requested):
Initial round of comments, just some nits

---

>Added an alarm test (2 seconds delay) after the time test. Is this enough ?

Yeah, should be fine


 >Look for a callback/gpio in function argument => create GPIO interrupt => call the callback function when interrupt triggered.
If no callback/gpio provided: don't try to create a GPIO interrupt.

Yes, as long as there's a separate function to initialize the interrupt support on the device, to initialize the GPIO in the correct input mode, etc.

>If no GPIO interrupt support: don't look for a callback. (I don't really know how to do this, I'm guessing some kind of pseudo-module as @jeandudey mentioned).

Possibly a `ds3231_int` pseudo-module that requires the `periph_gpio_irq` dependency, that way in the code you can decide if you provide or not the interrupt functionality. Pseudo-modules are added to the [makefiles/pseudomodules.inc.mk](https://github.com/RIOT-OS/RIOT/blob/master/makefiles/pseudomodules.inc.mk) file in alphabetical order, and the dependency on `periph_gpio_irq` is added on the `drivers/ds3231/Makefile.dep` file.

For example:

https://github.com/RIOT-OS/RIOT/blob/e768a85f6225b28ad9347debe36c01bb1f265825/drivers/itg320x/Makefile.dep#L4-L6

After that, you should then be able to use the C preprocessor to decide which code goes where, e.g.:

```c
#if IS_USED(MODULE_DS3231_INT)
int ds3231_init_int(ds3231_t *dev, ds3231_alarm_cb_t callback, void *arg)
{
    ...
}
#endif
```

nicoHarel commented:
Updated with ISR features, also previous comments have been taken into account, I hope I didn't forget anything. 

Test has also been updated based on what is with itg320x. 

Documentation is still missing for now.

jeandudey commented on drivers/ds3231/Makefile.dep:
```diff
@@ -1 +1,5 @@
+
+ifneq (,$(filter ds3231_int,$(USEMODULE))) 
+   FEATURES_REQUIRED += periph_gpio_irq 
+ endif
```
```suggestion
endif
```

jeandudey commented on drivers/ds3231/ds3231.c:
```diff
@@ -171,9 +176,25 @@ int ds3231_init(ds3231_t *dev, const ds3231_params_t *params)
+    assert(dev != NULL);
+    assert(gpio_is_valid(dev->int_pin));
+
+    /* SQW is set to low when an alarm is trigerred */
+    gpio_init_int(dev->int_pin, GPIO_IN, GPIO_FALLING, cb, arg);
```
```suggestion
    /* SQW is set to low when an alarm is triggered */
    if (gpio_init_int(dev->int_pin, GPIO_IN, GPIO_FALLING, cb, arg) < 0) {
        return -EIO;
    }
```

In the rare case that `gpio_init_int` fails 😉

jeandudey commented on drivers/include/ds3231.h:
```diff
@@ -53,11 +54,29 @@ enum {
 };
 
+typedef enum {
+    DS2321_AL1_TRIG_PER_S = 0x0F, /* alarm once per second */
+    DS2321_AL1_TRIG_S = 0x0E, /* alarm when second match */
+    DS2321_AL1_TRIG_M_S = 0x0C, /* alarm when minutes and seconds match */
+    DS2321_AL1_TRIG_H_M_S = 0x08, /* alarm when hours minutes seconds match */
+    DS2321_AL1_TRIG_D_H_M_S = 0x00, /* alarm when D/H/M/S match */
```
```suggestion
    DS2321_AL1_TRIG_PER_S = 0x0F, /**< alarm once per second */
    DS2321_AL1_TRIG_S = 0x0E, /**< alarm when second match */
    DS2321_AL1_TRIG_M_S = 0x0C, /**< alarm when minutes and seconds match */
    DS2321_AL1_TRIG_H_M_S = 0x08, /**< alarm when hours minutes seconds match */
    DS2321_AL1_TRIG_D_H_M_S = 0x00, /**< alarm when D/H/M/S match */
```

Applicable to the other enumerations too.

jeandudey commented on tests/driver_ds3231/main.c:
```diff
@@ -33,6 +33,8 @@
 
 static ds3231_t _dev;
 
+static kernel_pid_t p_main;
```
```suggestion
#ifdef MODULE_DS3231_INT
static kernel_pid_t p_main;
#endif
```

jeandudey commented on tests/driver_ds3231/main.c:
```diff
@@ -276,6 +334,8 @@ int main(void)
 {
     int res;
 
+    p_main = thread_getpid();
```
```suggestion
#ifdef MODULE_DS3231_INT
    p_main = thread_getpid();
#endif
```

jeandudey reviewed (commented):
Minor review comments.

Aside from the function documentations I see that most of the functionality is there, provided that docs are added and if anyone (or you) test that works showing the result this should be merged soon

nicoHarel commented:
Bit busy at the moment so I can only partially work on this, sorry if it is going a bit slowly. 

- I've corrected the issues you pointed out. 
- Added the documentation. 
- Corrected the enumerates name (DS2331 instead of DS3231) which apparently nobody included me ever noticed before :smile: 
- Removed clearing of previous alarm flags in "ds3231_set_alarm_1" and "ds3231_set_alarm_2"  as it' not really their job and there is now two dedicated functions for that.
- Added an option in device initialization so the interrupts can be enabled or disabled at startup
- I think that's it.

benpicco commented on drivers/ds3231/ds3231.c:
```diff
@@ -171,9 +176,32 @@ int ds3231_init(ds3231_t *dev, const ds3231_params_t *params)
 }
 
+#if IS_USED(MODULE_DS3231_INT)
+int ds3231_init_int(ds3231_t *dev, ds3231_alarm_cb_t cb, void *arg)
```
I'm a bit concerned with exposing the raw interrupt callback to the user.
We can't do I2C communication inside interrupt context, but the requirement to do `ds3231_get_alarm_1_flag` / `ds3231_clear_alarm_1_flag` could lead to just that.

How about 

```C
static void _unlock(void *m)
{
    mutex_unlock(m);
}

int ds3231_await_alarm(ds3231_t *dev)
{
    mutex_t mutex = MUTEX_INIT_LOCKED;

    if (gpio_init_int(dev->int_pin, GPIO_IN, GPIO_FALLING, _unlock, &mutex) < 0) {
            return -EIO;
    }

    /* wait for alarm */
    mutex_lock(&mutex);

    uint8_t status, tmp;
    _read(dev, REG_STATUS, &status, 1, 1, 0);

    /* clear interrupt flags */
    tmp = status & ~(STAT_A1F | STAT_A2F);
    res = _write(dev, REG_STATUS, &tmp, 1, 0, 1)
    if (res != 0) {
        return -EIO;
    }

    return status & (STAT_A1F | STAT_A2F);
}
```

nicoHarel replied:
This is interesting indeed. This literally happened to me a few days ago: tried to clear the flag inside the interrupt => got stuck waiting for a mutex that never came => wasted a few minutes trying to figure out what was going on. Therefore I can understand that it could be considered dangerous. 

Your proposition is fine by me, the only drawback that I see compared to a "user interrupt" is that you would have to block a thread waiting for the mutex while you wait for the alarm. But I'm guessing this is fine since in most cases, this is going to be the desired behavior anyway.

I won't modify the driver today in case someone disagrees with this.

nicoHarel replied:
OK so it seems everyone agrees with this implementation. 

Quick remark though @benpicco : do we want  `ds3231_await_alarm`to clear the alarm flag ? Is it its job ? User might want to clear it himself later (clearing the flag changes the logic state of the RTC alarm pin) for whatever reason(s). 

Just asking, it doesn't change anything for me other than how the alarm test is going to be implemented.

benpicco replied:
Is there a use case for not clearing the flag?
If the alarm pin is only used to signal the MCU, why would someone want to keep it high after the signal was delivered?

nicoHarel replied:
In case the alarm pin is not only attached to the MCU but to some other components/IC: syncing multiple MCUs, powering up some external sensors...

I admit those are some very exotic use cases :smiley:

nicoHarel replied:
Ok, I've updated the driver and the test method, also fixed some "static-tests" issues.

nicoHarel commented:
> Aside from the function documentations I see that most of the functionality is there, provided that docs are added and if anyone (or you) test that works showing the result this should be merged soon

How can I do that ? I mean, apart from showing the output of test sequence. Or maybe that's enough ?

jeandudey commented:
>How can I do that ? I mean, apart from showing the output of test sequence.

Yeah exactly 😉 , showing the output of the test sequence.

nicoHarel commented:
> > How can I do that ? I mean, apart from showing the output of test sequence.
> 
> Yeah exactly wink , showing the output of the test sequence.

No problem.

Ok so here are the results.

With the following lines in the makefile:
```
USEMODULE += ds3231_int
CFLAGS +="-DDS3231_PARAM_INT_PIN=(GPIO_PIN(1,2))"
```

```
2021-03-26 10:41:09,906 # main(): This is RIOT! (Version: 2021.04-devel-1028-g25c62-drivers/ds3231)
2021-03-26 10:41:09,907 # DS3231 RTC test
2021-03-26 10:41:09,907 # 
> test
2021-03-26 10:41:28,042 # test
2021-03-26 10:41:28,058 # testing device now
2021-03-26 10:41:32,052 # OK
```

With above lines in the makefile commented (no ds3231_int module):
```
2021-03-26 10:39:45,979 # main(): This is RIOT! (Version: 2021.04-devel-1028-g25c62-drivers/ds3231)
2021-03-26 10:39:45,980 # DS3231 RTC test
2021-03-26 10:39:45,980 # 
> test
2021-03-26 10:39:51,156 # test
2021-03-26 10:39:51,158 # testing device now
2021-03-26 10:39:55,182 # OK
```

jeandudey commented:
The changes look good to me, I don't know if @benpicco will want to take a look at this first :-)

benpicco commented on drivers/ds3231/ds3231.c:
```diff
@@ -236,6 +257,158 @@ int ds3231_set_time(const ds3231_t *dev, const struct tm *time)
+    }
+}
+
+int ds3231_clear_alarm_1_flag(const ds3231_t *dev)
+{
+    return _clrset(dev, REG_STATUS, STAT_A1F, 0, 1, 1);
+}
+
+int ds3231_clear_alarm_2_flag(const ds3231_t *dev)
+{
+    return _clrset(dev, REG_STATUS, STAT_A2F, 0, 1, 1);
+
+}
+
+int ds3231_get_alarm_1_flag(const ds3231_t *dev, bool *flag)
+{
+    uint8_t raw;
+    int res = _read(dev, REG_STATUS, &raw, 1, 1, 1);
+    if (res != 0) {
+        return -EIO;
+    }
+
+    *flag = (raw & STAT_A1F);
+
+    return res;
+}
+
+int ds3231_get_alarm_2_flag(const ds3231_t *dev, bool *flag)
+{
+    uint8_t raw;
+    int res = _read(dev, REG_STATUS, &raw, 1, 1, 1);
+    if (res != 0) {
+        return -EIO;
+    }
+
+    *flag = (raw & STAT_A2F) >> 1;
+
+    return res;
+}
```
Are those even needed anymore?

nicoHarel replied:
I needed those ds3231_clear_alarm_x_flag. Because, as I have explained previously, the alarm pin is not always attached to the MCU and can be used only on a voltage regulator (that's just an example based on my previous use case), sensors, another MCU...

An alarm doesn't necessary mean you want to "clean" it right away. In my instance, I wake up the MCU from deep sleep with its the reset pin (RTC alarm pin through a one shot IC). I want to release that alarm when I wake up and the alarm has already been triggered. Not while I wait for it, or not while I program another alarm (I might do that sometimes). 

The "ds3231_get_alarm_X_flag" function: how do you wait for an alarm if you don't have access to GPIO interrupt ?

benpicco commented on drivers/ds3231/ds3231.c:
```diff
@@ -236,6 +257,158 @@ int ds3231_set_time(const ds3231_t *dev, const struct tm *time)
+}
+#endif
+
+int ds3231_set_alarm_1(const ds3231_t *dev, struct tm *time,
+                       ds3231_alm_1_mode_t trigger)
```
```suggestion
typedef enum {
    DS3231_ALARM1,
    DS3231_ALARM2,
} ds3231_alarm_slot_t;

int ds3231_set_alarm(const ds3231_t *dev, ds3231_alarm_slot_t slot, struct tm *time,
                     ds3231_alm_1_mode_t trigger)
```

I think you can share some code here.

nicoHarel replied:
It would be nice but the triggers are different for the two alarms (see ds3231_alm_1_mode_t and ds3231_alm_2_mode_t typedef). I'm open to a different solution if someone finds one.

nicoHarel commented:
I haven't had news on this for a few days. @benpicco does that mean you agree with what I answered to you ? Can this be merged ? 

FYI: I won't be available for 2-3 weeks starting Thursday.

benpicco commented:
Sorry I was busy and didn't have time to look at this again yet.

benpicco commented on tests/driver_ds3231/main.c:
```diff
@@ -258,6 +262,70 @@ static int _cmd_test(int argc, char **argv)
+        puts("error: unable to program GPIO interrupt or to clear alarm flag");
+    }
+
+    if (!(res & 0x01)){
```
I'd prefer 0x1 to be a define (`DS3231_ALARM_FLAG_1` maybe)

nicoHarel replied:
Yes, a define is way better. Not sure where I was supposed to put it so I added it at the top of the test script.

benpicco commented on tests/driver_ds3231/main.c:
```diff
@@ -258,6 +262,70 @@ static int _cmd_test(int argc, char **argv)
+    puts("OK");
+    return 0;
+
+#endif
```
```suggestion
#else
```
to make it clearer that the other path is only take for the `!MODULE_DS3231_INT` case

benpicco commented on tests/driver_ds3231/main.c:
```diff
@@ -276,10 +344,17 @@ int main(void)
 {
     int res;
 
+#ifdef MODULE_DS3231_INT
+    p_main = thread_getpid();
+#endif
```
`p_main` is unused

benpicco commented on drivers/include/ds3231.h:
```diff
@@ -101,6 +153,96 @@ int ds3231_get_time(const ds3231_t *dev, struct tm *time);
+
+/**
+ * @brief   Get the state of alarm 1 flag (A1F)
+ *
```
Maybe add that this is not needed when `ds3231_await_alarm` is used, otherwise I see people use this function after `ds3231_await_alarm` and wonder why it returns `false`.

nicoHarel commented:
Updated. Thanks for the review :smiley:

benpicco commented on tests/driver_ds3231/main.c:
```diff
@@ -28,15 +28,13 @@
 #include "ds3231.h"
 #include "ds3231_params.h"
 
+#define DS3231_FLAG_ALARM_1 (0x01)
```
That should be in the driver header, along with one for `DS3231_FLAG_ALARM_2` and the information that it's returned by `ds3231_await_alarm`

nicoHarel replied:
Yes that makes a lot more sense.

benpicco commented:
Please squash

nicoHarel commented:
> Please squash

Done.

benpicco commented on drivers/ds3231/ds3231.c:
```diff
@@ -236,6 +257,158 @@ int ds3231_set_time(const ds3231_t *dev, const struct tm *time)
+
+    /* wait for alarm */
+    mutex_lock(&mutex);
+
```
You should also cal `gpio_irq_disable` here, otherwise if the interrupt happens a second time (when outside this function) `&mutex` will point to random stack memory.

nicoHarel replied:
Good catch. But quick question, does `gpio_init_int` always does some sort of a `gpio_irq_enable` (it looks like it but I'm not sure) ? Otherwise next time `ds3231_await_alarm` is called, the GPIO irq will be disabled if, I'm not mistaken.

nicoHarel commented:
Fixed the mutex issue. Squashed again. Thanks.

fjmolinas commented:
All green here @benpicco!
